files/check_keepalive.rb

```diff
@@ -0,0 +1,97 @@
+module Sensu
+module Extension
+
+# 1) update per-check timestamps in redis
+# 2) have eventmachine periodic timer to check if anything timed out and
+#    generate a new "critical" event for every
+#
+# extension has access to:
+#  - logger     @logger
+#  - redis      @settings[:redis]
+#  - transport  @settings[:transport]
+
+class CheckKeepalive < Handler
+  def name
+    'check_keepalive'
+  end
+
+  def description
+    'emit per-check keepalive events'
+  end
+
+  def post_init
+    @redis    = @settings.redis
+    @em_timer = EventMachine::PeriodicTimer.new(10) { process_keepalives }
+  end
+
+  def stop
+    @em_timer.cancel
+    yield
+  end
+
+  def run(event)
+    client_name = event[:client][:name]
+    check_name  = event[:check][:name]
+
+    rkey = "ext:check_keepalive:checks"
```

Comment:
When a check is commented out, puppet removes the file and we have a special notify exec that resolves the check if it exists. If not, no big deal.

Again, check keepalives are really not needed on client-scheduled events. On client-scheduled events (puppet), the client is responsible for making sure checks continue to run. On non-client-scheduled events, we (I) want the *server* responsible for making sure check results come in, and hence the need for this extension.

We only need keepalive data on non-puppet deployed checks.